### PR TITLE
Fixes #175 Liberty tool window doesn't scroll

### LINKED ISSUE #175: Liberty tool window doesn't scroll
When the Liberty tool window is shrunk too small to show everything, there is no scrolling/scrollbar. The user must vertically expand the window to view/access the hidden commands.
image

### PR DESCRIPTION
**Pull request by anusreelakshmi934**
Fixes #175
Fixed the issue of Liberty tool window doesn't scroll.

### CODE REVIEW

**Comment by TrevCraw** [comment]
It looks like these changes affect the automated tests. Please update the tests to accommodate the changes.

Test failures are occurring on these lines in the log:
https://github.com/OpenLiberty/liberty-tools-intellij/actions/runs/6047671095/job/16411560760?pr=485#step:6:8604
https://github.com/OpenLiberty/liberty-tools-intellij/actions/runs/6047671095/job/16411560760?pr=485#step:6:8846
https://github.com/OpenLiberty/liberty-tools-intellij/actions/runs/6047671095/job/16411560760?pr=485#step:6:9012
https://github.com/OpenLiberty/liberty-tools-intellij/actions/runs/6047671095/job/16411560760?pr=485#step:6:9276
https://github.com/OpenLiberty/liberty-tools-intellij/actions/runs/6047671095/job/16411560760?pr=485#step:6:9481

File: src/main/java/io/openliberty/tools/intellij/actions/RefreshLibertyToolbar.java [hunk 1]
```diff
@@ -60,8 +61,9 @@ public static void refreshDashboard(Project project) {
         Component existingTree = null;
         Component existingActionToolbar = null;
         for (Component comp : components) {
-            if (comp.getName() != null && comp.getName().equals(Constants.LIBERTY_TREE)) {
-                existingTree = comp;
+            if (comp.getName() != null && comp.getName().equals(Constants.LIBERTY_SCROLL_PANE)) {
+                JBScrollPane scrollPane = (JBScrollPane) comp;
```
Comment (aparnamichael) [inline src/main/java/io/openliberty/tools/intellij/actions/RefreshLibertyToolbar.java:65]: Before doing typecasting it is always better to check
if (comp instanceof JBScrollPane) / if (scrollPane.getViewport().getView() instanceof Tree)

else there is a chance of JavaTypecastException.

File: src/main/java/io/openliberty/tools/intellij/actions/RunLibertyDevTask.java [hunk 2]
```diff
@@ -57,8 +58,9 @@ public void update(@NotNull AnActionEvent e) {
             JComponent libertyWindow = content.getComponent();
             Component[] components = libertyWindow.getComponents();
             for (Component comp : components) {
-                if (comp.getName() != null && comp.getName().equals(Constants.LIBERTY_TREE)) {
-                    Tree libertyTree = (Tree) comp;
+                if (comp instanceof JBScrollPane && comp.getName() != null && comp.getName().equals(Constants.LIBERTY_SCROLL_PANE)) {
+                    JBScrollPane scrollPane = (JBScrollPane) comp;
+                    Tree libertyTree = (Tree) scrollPane.getViewport().getView();
```
Comment (TrevCraw) [inline src/main/java/io/openliberty/tools/intellij/actions/RunLibertyDevTask.java:63]: As Aparna recommended in her review, it would be best to check the type of `scrollPane.getViewport().getView()` before typecasting as a `Tree`.

File: src/main/java/io/openliberty/tools/intellij/actions/RunLibertyDevTask.java [hunk 3]
```diff
@@ -104,8 +106,9 @@ public void actionPerformed(@NotNull AnActionEvent e) {
             Component[] components = libertyWindow.getComponents();
 
             for (Component comp : components) {
-                if (comp.getName() != null && comp.getName().equals(Constants.LIBERTY_TREE)) {
-                    Tree libertyTree = (Tree) comp;
+                if (comp instanceof JBScrollPane && comp.getName() != null && comp.getName().equals(Constants.LIBERTY_SCROLL_PANE)) {
+                    JBScrollPane scrollPane = (JBScrollPane) comp;
+                    Tree libertyTree = (Tree) scrollPane.getViewport().getView();
```
Comment (TrevCraw) [inline src/main/java/io/openliberty/tools/intellij/actions/RunLibertyDevTask.java:111]: Same comment as above.

**Review by TrevCraw — CHANGES_REQUESTED** [review]

File: src/main/java/io/openliberty/tools/intellij/actions/RunLibertyDevTask.java [hunk 4]
```diff
@@ -104,8 +107,10 @@ public void actionPerformed(@NotNull AnActionEvent e) {
             Component[] components = libertyWindow.getComponents();
 
             for (Component comp : components) {
-                if (comp.getName() != null && comp.getName().equals(Constants.LIBERTY_TREE)) {
-                    Tree libertyTree = (Tree) comp;
+                if (comp instanceof JBScrollPane && comp.getName() != null && comp.getName().equals(Constants.LIBERTY_SCROLL_PANE)) {
+                    JBScrollPane scrollPane = (JBScrollPane) comp;
+                    if (scrollPane.getViewport().getView() instanceof Tree) {
+                        Tree libertyTree = (Tree) scrollPane.getViewport().getView();
```
Comment (TrevCraw) [inline src/main/java/io/openliberty/tools/intellij/actions/RunLibertyDevTask.java:113]: The other opportunity for improvement for this code has to do with the fact that there is code duplication between the `update` and `actionPerformed` methods (once again, I know this code was already here and not something you wrote). We should pull out the duplicate code and put it in a separate method or methods that we can call from `update` and from `actionPerformed`.

The duplication is between this section: https://github.com/anusreelakshmi934/liberty-tools-intellij/blob/bugfix/%23175/src/main/java/io/openliberty/tools/intellij/actions/RunLibertyDevTask.java#L55-L79   
...and this section: https://github.com/anusreelakshmi934/liberty-tools-intellij/blob/bugfix/%23175/src/main/java/io/openliberty/tools/intellij/actions/RunLibertyDevTask.java#L101-L126

Please find the duplicate code and separate it into methods that can be called from `update` and `actionPerformed`.

**Review by TrevCraw — CHANGES_REQUESTED** [review]

File: src/main/java/io/openliberty/tools/intellij/actions/RunLibertyDevTask.java (hunk 4 above)
Comment (TrevCraw) [reply src/main/java/io/openliberty/tools/intellij/actions/RunLibertyDevTask.java]: @anusreelakshmi934 the common method you created is very good. We can go ahead and keep it as is. I have just a few tips for improvement in the future. 

Typically, when creating common methods, you want to make it so they can be reused without having to modify them in the future. With the way that your common method is structured, if a new method is written that needs to handle the incoming "Action Event" on the Liberty Tree in a new way, your common method will need to be updated to handle the new scenario. This means the `boolean` parameter `isUpdate` won't be useful anymore (since there is now a third option) and new code will need to be added in the section after retrieving the `lastPathComponent`. 

It would be a better design to keep the unique parts of `update` and `actionPerformed` as part of those methods and set up the common method (or methods) so that it does not need to be modified for new methods to make use of it. Just something to keep in mind moving forward.

**Review by TrevCraw — CHANGES_REQUESTED** [review]

File: src/main/java/io/openliberty/tools/intellij/actions/RunLibertyDevTask.java [hunk 5]
```diff
@@ -68,13 +68,14 @@ private static void getLibertyTree(@NotNull AnActionEvent e, Project project, bo
                     if (view instanceof Tree) {
                         libertyTree = (Tree) view;
                         TreePath[] selectionPaths = libertyTree.getSelectionPaths();
-                        // when only one child node is selected, enable this action
                         if (selectionPaths != null && selectionPaths.length == 1) {
                             String lastPathComponent = selectionPaths[0].getLastPathComponent().toString();
                             if (Constants.FULL_ACTIONS_MAP.containsKey(lastPathComponent)) {
                                 if (isUpdate) {
                                     e.getPresentation().setEnabled(true);
+                                    // when only one child node is selected, enable this action
```
Comment (TrevCraw) [inline src/main/java/io/openliberty/tools/intellij/actions/RunLibertyDevTask.java:76]: It is standard for comments to apply to the lines they are on or the lines below. It may be misleading to have this comment after the line it is meant to be about.

File: src/main/java/io/openliberty/tools/intellij/lsp4jakarta/lsp4ij/codeAction/proposal/quickfix/RemoveParamAnnotationQuickFix.java [hunk 6]
```diff
@@ -1,5 +1,5 @@
  /*******************************************************************************
- * Copyright (c) 2021, 2023 IBM Corporation and others.
+ * Copyright (c) 2021, 2024 IBM Corporation and others.
```
Comment (TrevCraw) [inline src/main/java/io/openliberty/tools/intellij/lsp4jakarta/lsp4ij/codeAction/proposal/quickfix/RemoveParamAnnotationQuickFix.java:2]: not sure why this copyright is being updated - this file is not related to this PR

**Review by dessina-devasia — APPROVED** [review]
Looks good for me

File: src/main/java/io/openliberty/tools/intellij/actions/RunLibertyDevTask.java [hunk 7]
```diff
@@ -51,28 +52,41 @@ public void update(@NotNull AnActionEvent e) {
             //if the action event data context returns a null project, just return and let successive update calls modify the presentation
             return;
         }
+        handleLibertyTreeEvent(e, project, true);
+    }
+
+    private static void handleLibertyTreeEvent(@NotNull AnActionEvent e, Project project, boolean isUpdate) {
```
Comment (TrevCraw) [inline src/main/java/io/openliberty/tools/intellij/actions/RunLibertyDevTask.java:58]: Is there a reason this method was made `static`? It does not seem like it needs to be.

Comment (anusreelakshmi934) [reply src/main/java/io/openliberty/tools/intellij/actions/RunLibertyDevTask.java:58]: No, it was unnecessary. I've already implemented the modification.

**Review by TrevCraw — APPROVED** [review]

**Review by dessina-devasia — APPROVED** [review]
